addons/mail/models/mail_activity.py

```diff
@@ -571,53 +584,129 @@ def action_open_document(self):
+                                    for activity in completed_activities if activity.get('attachment_ids')
+                                    for attachment_id in activity['attachment_ids']]
+            attachments_info = {}
+            if activity_attachments:
+                last_attachment = max(activity_attachments, key=lambda a: a['create_date'])
+                attachments_info['attachments'] = {
+                    'last': {
+                        'id': last_attachment['id'],
+                        'name': last_attachment['name']
+                    },
+                    'count': len(activity_attachments),
+                }
+            grouped_activities[res_id][activity_type_id] = {
+                'count_by_state': {
+                    **dict(Counter(map(lambda a: self._compute_state_from_date(a['date_deadline'], user_tz),
+                                       ongoing_activities))),
+                    **({'done': len(completed_activities)} if completed_activities else {}),
+                },
+                'ids': [a['id'] for a in ongoing_activities] + [a['id'] for a in completed_activities],
+                'o_closest_date': date_done if is_all_activities_done else date_deadline,
+                'state': 'done' if is_all_activities_done else self._compute_state_from_date(date_deadline, user_tz),
+                'user_ids_ordered_by_deadline': distinct_assignees,
+                **attachments_info
             }
+
         activity_type_infos = []
-        activity_type_ids = self.env['mail.activity.type'].search(
-            ['|', ('res_model', '=', res_model), ('res_model', '=', False)])
-        for elem in sorted(activity_type_ids, key=lambda item: item.sequence):
+        for activity_type in sorted(activity_types, key=lambda item: item.sequence):
             mail_template_info = []
-            for mail_template_id in elem.mail_template_ids:
+            for mail_template_id in activity_type.mail_template_ids:
                 mail_template_info.append({"id": mail_template_id.id, "name": mail_template_id.name})
-            activity_type_infos.append([elem.id, elem.name, mail_template_info])
+            activity_type_infos.append([activity_type.id, activity_type.name, mail_template_info,
+                                        activity_type.keep_done])
 
         return {
             'activity_types': activity_type_infos,
-            'activity_res_ids': sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item]),
-            'grouped_activities': activity_data,
+            'activity_res_ids': (  # record with ongoing activities followed by the ones with all completed activities
+                    sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item]) +
+                    sorted({k: v for k, v in res_id_to_date_done.items() if k not in res_id_to_deadline},
+                           key=lambda item: res_id_to_date_done[item], reverse=True)
```

Comment:
is it really necessary ? can't we do just (test are green with that)
```py
sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item])
```

(because if the dateline is set, datedone will always be set)

Reply:
and so the variable `res_id_to_date_done` is now useless